merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/CheckpointSimulationDriver.java

```diff
@@ -0,0 +1,412 @@
+          combinedTimeline.add(t.events());
+        }
+      }
+    }
+
+    for (final var timePoint : timeline) {
+      if (timePoint instanceof TemporalEventSource.TimePoint.Delta t) {
+        combinedTimeline.add(t.delta());
+      } else if (timePoint instanceof TemporalEventSource.TimePoint.Commit t) {
+        combinedTimeline.add(t.events());
+      }
+    }
+    return combinedTimeline;
+  }
+
+  public static Function<SimulationState, Boolean> desiredCheckpoints(final List<Duration> desiredCheckpoints) {
+    return simulationState -> {
+      for (final var desiredCheckpoint : desiredCheckpoints) {
+        if (simulationState.currentTime().noLongerThan(desiredCheckpoint) && simulationState.nextTime().longerThan(desiredCheckpoint)) {
+          return true;
+        }
+      }
+      return false;
+    };
+  }
+
+  public static Function<SimulationState, Boolean> wallClockCheckpoints(final long thresholdSeconds) {
+    MutableLong lastCheckpointRealTime = new MutableLong(System.nanoTime());
+    MutableObject<Duration> lastCheckpointSimTime = new MutableObject<>(Duration.ZERO);
+    return simulationState -> {
+      if (simulationState.nextTime().longerThan(simulationState.currentTime()) && System.nanoTime() - lastCheckpointRealTime.getValue() > (thresholdSeconds * 1000 * 1000 * 1000)) {
+        lastCheckpointRealTime.setValue(System.nanoTime());
+        lastCheckpointSimTime.setValue(simulationState.currentTime());
+        return true;
+      } else {
+        return false;
+      }
+    };
+  }
+
+  public static Function<SimulationState, Boolean> checkpointAtEnd(Function<SimulationState, Boolean> stoppingCondition) {
+    return simulationState -> stoppingCondition.apply(simulationState) || simulationState.nextTime.isEqualTo(MAX_VALUE);
+  }
+
+  public record SimulationState(
+      Duration currentTime,
+      Duration nextTime,
+      SimulationEngine simulationEngine,
+      Map<ActivityDirectiveId,  ActivityDirective> schedule,
+      Map<ActivityDirectiveId, SpanId> activityDirectiveIdSpanIdMap
+  ){}
+
+  public static <Model> SimulationResultsComputerInputs simulateWithCheckpoints(
+      final MissionModel<Model> missionModel,
+      final Map<ActivityDirectiveId, ActivityDirective> schedule,
+      final Instant simulationStartTime,
+      final Duration simulationDuration,
+      final Instant planStartTime,
+      final Duration planDuration,
+      final Consumer<Duration> simulationExtentConsumer,
+      final Supplier<Boolean> simulationCanceled,
+      final CachedSimulationEngine cachedEngine,
+      final Function<SimulationState, Boolean> shouldTakeCheckpoint,
+      final Function<SimulationState, Boolean> stopConditionOnPlan,
+      final CachedEngineStore cachedEngineStore,
+      final SimulationEngineConfiguration configuration,
+      final boolean duplicationIsOk) {
+    final var activityToSpan = new HashMap<ActivityDirectiveId, SpanId>();
```

Comment:
Nitpick, non-blocking (applies to all instances throughout the file but I didn't want to add like 10 comments for this): 
```suggestion
      final boolean duplicationIsOk
  ) {
    final var activityToSpan = new HashMap<ActivityDirectiveId, SpanId>();
```

Reply:
I'd rather not.